miniapps/nurbs/nurbs_solenoidal.cpp

```diff
@@ -0,0 +1,398 @@
+   MixedBilinearForm *bVarf(new MixedBilinearForm(R_space, W_space));
+
+   mVarf->AddDomainIntegrator(new VectorFEMassIntegrator(one));
+   mVarf->Assemble();
+   mVarf->Finalize();
+
+   bVarf->AddDomainIntegrator(new VectorFEDivergenceIntegrator);
+   bVarf->Assemble();
+   bVarf->Finalize();
+
+   SparseMatrix &M(mVarf->SpMat());
+   SparseMatrix &B(bVarf->SpMat());
+   B *= -1.;
+   TransposeOperator *Bt = new TransposeOperator(&B);
+
+   BlockOperator darcyOp(block_offsets);
+   darcyOp.SetBlock(0,0, &M);
+   if (div_free) { darcyOp.SetBlock(0,1, Bt); }
+   if (div_free) { darcyOp.SetBlock(1,0, &B); }
+
+   // 10. Construct the operators for preconditioner
+   //
+   //                 P = [ diag(M)         0         ]
+   //                     [  0       B diag(M)^-1 B^T ]
+   //
+   //     Here we use Symmetric Gauss-Seidel to approximate the inverse of the
+   //     pressure Schur Complement
+   Vector Md(mVarf->Height());
+   M.GetDiag(Md);
+   Md.HostReadWrite();
+   SparseMatrix *MinvBt = Transpose(B);
+   for (int i = 0; i < Md.Size(); i++)
+   {
+      MinvBt->ScaleRow(i, 1./Md(i));
+   }
+   SparseMatrix *S = Mult(B, *MinvBt);
+   Solver *invS;
+#ifndef MFEM_USE_SUITESPARSE
+   invS = new GSSmoother(*S);
+#else
+   invS = new UMFPackSolver(*S);
+#endif
+   invS->iterative_mode = false;
+
+   Solver *invM = new GSSmoother(M);
+   invM->iterative_mode = false;
+
+   BlockDiagonalPreconditioner darcyPrec(block_offsets);
+   darcyPrec.SetDiagonalBlock(0, invM);
+   darcyPrec.SetDiagonalBlock(1, invS);
+
+   // 11. Solve the linear system with MINRES.
+   //     Check the norm of the unpreconditioned residual.
+   int maxIter(10000);
+   double rtol(1.e-16);
+   double atol(1.e-16);
+
+   chrono.Clear();
+   chrono.Start();
+   MINRESSolver solver;
+   solver.SetAbsTol(atol);
+   solver.SetRelTol(rtol);
+   solver.SetMaxIter(maxIter);
+   solver.SetOperator(darcyOp);
+   solver.SetPreconditioner(darcyPrec);
+   solver.SetPrintLevel(2);
+   x = 0.0;
+   solver.Mult(rhs, x);
+   if (device.IsEnabled()) { x.HostRead(); }
+   chrono.Stop();
+
+   if (solver.GetConverged())
+   {
+      std::cout << "MINRES converged in " << solver.GetNumIterations()
+                << " iterations with a residual norm of "
+                << solver.GetFinalNorm() << ".\n";
+   }
+   else
+   {
+      std::cout << "MINRES did not converge in " << solver.GetNumIterations()
+                << " iterations. Residual norm is " << solver.GetFinalNorm()
+                << ".\n";
+   }
+   std::cout << "MINRES solver took " << chrono.RealTime() << "s.\n";
+
+   // 12. Create the grid functions u and p
+   GridFunction u, p, uu, vv, ww;
+   u.MakeRef(R_space, x.GetBlock(0), 0);
+   p.MakeRef(W_space, x.GetBlock(1), 0);
+
+   // 13. Save the mesh and the solution. This output can be viewed later using
+   //     GLVis: "glvis -m ex5.mesh -g sol_u.gf" or "glvis -m ex5.mesh -g
+   //     sol_p.gf".
+   {
+      ofstream mesh_ofs("ex5.mesh");
```

Comment:
`ex5` -> `exsol`?

Reply:
Indeed! Done.

Reply:
I also update the makefile to make sure this input is removed when requested.